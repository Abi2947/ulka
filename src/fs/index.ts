export { default as allFiles } from "./all-files"
export { default as copyAssets } from "./copy-assets"
export { default as mkdir } from "./mkdir"
export { default as rmdir } from "./rmdir"
