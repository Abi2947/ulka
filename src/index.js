const { generateHash, getConfigs } = require("./utils/helpers")
const log = require("./utils/ulka-log")

module.exports = {
  generateHash,
  getConfigs,
  log
}
