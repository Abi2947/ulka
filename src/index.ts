export { default as UlkaSource, UlkaSourceContext } from "./source/ulka-source"
export { default as MdSource, MDSourceContext } from "./source/md-source"
export { default as Source, SourceContext } from "./source"
export {
  PluginBeforeMd,
  PluginAfterMd,
  PluginAfterUlka,
  PluginBeforeUlka,
  UlkaPluginRehype,
  UlkaPluginRemark
} from "./data/plugins"
