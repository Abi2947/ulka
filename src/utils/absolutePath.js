const path = require('path')

const absolutePath = args => {
  if (typeof args !== 'string') {
    throw new Error('Path provided should be string')
  }
  return path.join(process.cwd(), ...args.split('/'))
}

module.exports = absolutePath
