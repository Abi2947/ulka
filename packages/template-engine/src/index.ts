import fs from 'fs'
import vm from 'vm'

export interface Options {
  base: string
}

export { render }

function render(str: string, data: { [key: string]: any }, options: Options) {
  const ctx = vm.createContext(context(data, options))

  return str.replace(/\\?{%(.*?)%}/gs, (match, js, index) => {
    if (match[0] === '\\' && str[index - 1] !== '\\') {
      return match.slice(1)
    }

    let result = vm.runInContext(js, ctx)

    if (Array.isArray(result)) {
      result = result.join('')
    }

    return result || ''
  })
}

function context(ctx: any, options: Options) {
  return {
    ...ctx,
    require: (path: string) => customRequire(path, options),
    console,
  }
}

function customRequire(requirePath: string, options: Options) {
  if (!options.base) throw new Error(`"base" is a required option.`)

  const req = require.resolve(requirePath, { paths: [options.base] })

  try {
    return require(req)
  } catch (e) {
    if (fs.existsSync(req)) return fs.readFileSync(req, 'utf-8')
    else throw new Error(`Couldn't find file ${requirePath}`)
  }
}
