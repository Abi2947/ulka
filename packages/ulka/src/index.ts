import c from 'ansi-colors'
import rimraf from 'rimraf'

import type { FSWatcher } from 'chokidar'

import { Ulka } from './Ulka'
import { runPlugins, createWatcher, clearConsole } from './utils'

export async function setup(cwd: string, task: string, cpath: string) {
  return await new Ulka(cwd, task, cpath).setup()
}

export async function build(ulka: Ulka) {
  await runPlugins('beforeBuild', { ulka })

  try {
    await ulka.getLayouts()
    await ulka.getCollections()
    await ulka.write()
  } catch (e) {
    console.log('')
    console.log(e.message)
    e.custom && console.log(c.redBright(`> ${e.custom}`))
  }

  await runPlugins('afterBuild', { ulka })
}

export async function watch(
  ulka: Ulka,
  verbose = false,
  customFn?: Function
): Promise<FSWatcher> {
  await build(ulka)

  const watcher = createWatcher(ulka)

  watcher.on('all', async (eventName, filepath) => {
    if (!['add', 'change', 'unlink'].includes(eventName)) return

    clearConsole()
    console.log(c.blueBright(`> Change detected in ${filepath}\n`))

    eventName === 'unlink' &&
      rimraf.sync(ulka.configs.output, { disableGlob: true })

    ulka.reset()
    ulka.configs.verbose = verbose
    await build(ulka)

    ulka.server.log()

    filepath.endsWith('.css')
      ? ulka.server.wss.reloadCss()
      : ulka.server.wss.reload()

    customFn && customFn(watcher)
  })

  return watcher
}

export { Ulka, clearConsole }
export { FileInfo } from './FileInfo'
export { defineConfig, box, resolvePlugin } from './utils'
export { UlkaError } from './UlkaError'
export { Collection } from './Collection'
export { UlkaServer } from './UlkaServer'
export type {
  Configs,
  ContentConfig,
  PluginConfig,
  Plugins,
  PluginFunction,
  ValidContentConfig,
  PluginArg,
  PluginName,
} from './types'
export {
  Engines,
  EjsTemplate,
  LiquidTemplate,
  MdTemplate,
  Template,
  UlkaTemplate,
} from './Templates'
