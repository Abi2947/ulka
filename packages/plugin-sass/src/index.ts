import fs from 'fs'
import sass from 'sass'
import { Template, Plugin } from 'ulka'

class SassTemplate extends Template {
  static sass = sass
  static opts: sass.Options = {}

  async compile() {
    return async () => {
      const result = SassTemplate.sass.renderSync({
        data: this.content as string,
        file: this.fileinfo.filepath,
        outFile: this.buildPath,
        ...SassTemplate.opts,
        indentedSyntax: this.fileinfo.parsedpath.ext === '.sass',
      })

      result.map && fs.writeFileSync(this.buildPath + '.map', result.map)

      return result.css
    }
  }

  get hasMatter() {
    return false
  }

  get buildExt() {
    return '.css'
  }

  get contentShouldBeString() {
    return true
  }
}

const plugin: Plugin = (opts: sass.Options = {}) => ({
  afterSetup({ ulka }) {
    SassTemplate.opts = opts

    ulka.engines['.sass'] = SassTemplate
    ulka.engines['.scss'] = SassTemplate
  },
})

export default plugin
