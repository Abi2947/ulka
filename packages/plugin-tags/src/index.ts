import { PluginFunction } from 'ulka'

type PluginType = (options?: { matterKey?: string; contextKey?: string }) => {
  afterCreateContext: PluginFunction<'afterCreateContext'>
  afterBuild: PluginFunction<'afterBuild'>
}

const plugin: PluginType = ({
  matterKey = 'tags',
  contextKey = 'tags',
} = {}) => {
  let tags: { [key: string]: any } = {}

  return {
    afterCreateContext: ({ content }) => {
      content.context[contextKey] = tags

      if (content.context.matter[matterKey]) {
        for (const tag of content.context.matter[matterKey]) {
          if (!tags[tag]) tags[tag] = []

          tags[tag].push(content.context)
        }
      }
    },
    afterBuild() {
      tags = {}
    },
  }
}

export default plugin
