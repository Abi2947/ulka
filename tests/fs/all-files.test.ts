import path from "path"
import { allFiles } from "../../src/fs"
import absolutePath from "../../src/utils/absolute-path"

const testDir = "tests/resources/fs_test/allfiles"

describe("all Files function", () => {
  test("should return all files in a folder", () => {
    const allFilesList = allFiles(absolutePath(testDir)).map(f =>
      path.relative(process.cwd(), f).split(path.sep).join("/")
    )
    expect(allFilesList).toMatchSnapshot()
  })

  test("should return ulka files only", () => {
    const allUlkaFiles = allFiles(absolutePath(testDir), ".ulka").map(f =>
      path.relative(process.cwd(), f).split(path.sep).join("/")
    )
    expect(allUlkaFiles).toMatchSnapshot()
  })

  test("should return js files only", () => {
    const allJsFiles = allFiles(absolutePath(testDir), ".js").map(f =>
      path.relative(process.cwd(), f).split(path.sep).join("/")
    )
    expect(allJsFiles).toMatchSnapshot()
  })

  test("should return the file in a array if file path is provided", () => {
    const file1 = allFiles(absolutePath(testDir + "/file1.js")).map(f =>
      path.relative(process.cwd(), f).split(path.sep).join("/")
    )
    expect(file1).toMatchInlineSnapshot(`
      Array [
        "tests/resources/fs_test/allfiles/file1.js",
      ]
    `)
  })
})
