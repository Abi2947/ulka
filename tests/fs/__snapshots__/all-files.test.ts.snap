// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`all Files function should return all files in a folder 1`] = `
Array [
  "tests/resources/fs_test/allfiles/file1.js",
  "tests/resources/fs_test/allfiles/file1.ulka",
  "tests/resources/fs_test/allfiles/file2.js",
  "tests/resources/fs_test/allfiles/file2.ulka",
  "tests/resources/fs_test/allfiles/folder1/folder1-file1.ulka",
  "tests/resources/fs_test/allfiles/folder1/folder1-file2.ulka",
]
`;

exports[`all Files function should return js files only 1`] = `
Array [
  "tests/resources/fs_test/allfiles/file1.js",
  "tests/resources/fs_test/allfiles/file2.js",
]
`;

exports[`all Files function should return ulka files only 1`] = `
Array [
  "tests/resources/fs_test/allfiles/file1.ulka",
  "tests/resources/fs_test/allfiles/file2.ulka",
  "tests/resources/fs_test/allfiles/folder1/folder1-file1.ulka",
  "tests/resources/fs_test/allfiles/folder1/folder1-file2.ulka",
]
`;
